File: DAA_LAB/Experiment_4_CountingSort/CountingSort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void countingSort(int arr[], int n)
{
    int count[101] = {0};
    int output[n];

    for(int i = 0; i < n; i++)
        count[arr[i]]++;

    for(int i = 1; i <= 100; i++)
        count[i] += count[i - 1];

    for(int i = n - 1; i >= 0; i--)
    {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }

    for(int i = 0; i < n; i++)
        arr[i] = output[i];
}

int main()
{
    int n = 1000;

    int marks[n];

    srand(time(NULL));

    for(int i = 0; i < n; i++)
        marks[i] = rand() % 101;

    clock_t start = clock();

    countingSort(marks, n);

    clock_t end = clock();

    double executionTime =
        (double)(end - start) /
        CLOCKS_PER_SEC;

    printf("First 20 Sorted Marks:\n");

    for(int i = 0; i < 20; i++)
        printf("%d ", marks[i]);

    printf("\n");

    printf("Execution Time = %lf seconds\n",
            executionTime);

    return 0;
}

File: DAA_LAB/Experiment_4_CountingSort/CountingSort.py
import random
import time

def counting_sort(arr):

    count = [0] * 101

    output = [0] * len(arr)

    for num in arr:
        count[num] += 1

    for i in range(1, 101):
        count[i] += count[i - 1]

    for i in range(len(arr) - 1, -1, -1):

        output[count[arr[i]] - 1] = arr[i]

        count[arr[i]] -= 1

    return output


n = 1000

marks = [random.randint(0, 100)
         for _ in range(n)]

start = time.time()

marks = counting_sort(marks)

end = time.time()

print("First 20 Sorted Marks:")
print(marks[:20])

print("Execution Time =",
      end - start,
      "seconds")
